spec/controllers/samples_controller_spec.rb

```diff
@@ -0,0 +1,112 @@
+require 'rails_helper'
+
+RSpec.describe SamplesController, type: :controller do
+  create_users
+
+  pipeline_run_stages_data = [{
+    name: "Host Filtering",
+    dag_json: "{\"key1\": \"value1\"}"
+  }, {
+    name: "GSNAPL/RAPSEARCH alignment",
+    dag_json: "{\"key2\": \"value2\"}"
+  }, {
+    name: "Post Processing",
+    dag_json: "{\"key3\": \"value3\"}"
+  }, {
+    name: "Experimental",
+    dag_json: "{\"key4\": \"value4\"}"
+  }]
+
+  # Admin specific behavior
+  context "Admin user" do
+    before do
+      sign_in @admin
+      @admin.add_allowed_feature("pipeline_viz")
+      @admin.save!
+    end
+
+    describe "GET pipeline stage results" do
+      it "sees all pipeline stage results" do
+        project = create(:public_project)
+        sample = create(:sample, project: project,
+                                 pipeline_runs_data: [{ pipeline_run_stages_data: pipeline_run_stages_data }])
+
+        get :stage_results, params: { id: sample.id }
+
+        json_response = JSON.parse(response.body)["pipeline_stage_results"]
+        expect(json_response.length).to eq(4)
+        expect(json_response["Experimental"]).to be_truthy
+        pipeline_run_stages_data.each do |stage|
+          expect(json_response[stage[:name]]).to eq(JSON.parse(stage[:dag_json]))
+        end
+      end
+    end
+
+    describe "GET pipeline stage results without pipeline viz flag enabled" do
+      it "cannot see stage results" do
+        project = create(:public_project)
+        sample = create(:sample, project: project,
+                                 pipeline_runs_data: [{ pipeline_run_stages_data: pipeline_run_stages_data }])
+
+        @admin.remove_allowed_feature("pipeline_viz")
+        @admin.save!
+        get :stage_results, params: { id: sample.id }
+
+        expect(response).to have_http_status 401
+      end
+    end
+  end
+
+  # Non-admin, aka Joe, specific behavior
+  context "Joe" do
+    before do
+      sign_in @joe
+      @joe.add_allowed_feature("pipeline_viz")
+      @joe.save!
+    end
+
+    describe "GET pipeline stage results" do
+      it "can see pipeline stage results without the experimental stage results" do
+        project = create(:public_project)
```

Comment:
You do not have a test for a project of the user, right? I.e.
`create(:project, users:[@joe])`

Reply:
Added! https://github.com/chanzuckerberg/idseq-web/pull/2304/files#diff-40e83e0600ff9945e686badfda7541a1R88